xarray_beam/_src/pangeo_forge.py

```diff
@@ -0,0 +1,155 @@
+  return {dim: 0 for dim in dataset.dims.keys()}
+
+
+def _expand_dimensions_by_key(
+    dataset: xarray.Dataset,
+    index: 'FilePatternIndex',
+    pattern: 'FilePattern'
+) -> xarray.Dataset:
+  """Expand the dimensions of the `Dataset` by offsets found in the `Key`."""
+  combine_dims_by_name = {
+    combine_dim.name: combine_dim for combine_dim in pattern.combine_dims
+  }
+  index_by_name = {
+    idx.name: idx for idx in index
+  }
+
+  if not combine_dims_by_name:
+    return dataset
+
+  for dim_key in index_by_name.keys():
+    # skip expanding dimensions if they already exist
+    if dim_key in dataset.dims:
+      continue
+
+    try:
+      combine_dim = combine_dims_by_name[dim_key]
+    except KeyError:
+      raise ValueError(
+        f"could not find CombineDim named {dim_key!r} in pattern {pattern!r}."
+      )
+
+    dim_val = combine_dim.keys[index_by_name[dim_key].index]
+    dataset = dataset.expand_dims(**{dim_key: [dim_val]})
+
+  return dataset
+
+
+class FilePatternToChunks(beam.PTransform):
+  """Open data described by a Pangeo-Forge `FilePattern` into keyed chunks."""
+
+  from pangeo_forge_recipes.patterns import FilePattern, FilePatternIndex
+
+  def __init__(
+      self,
+      pattern: 'FilePattern',
+      chunks: Optional[Mapping[str, int]] = None,
+      local_copy: bool = True,
+      xarray_open_kwargs: Optional[Dict] = None
+  ):
+    """Initialize FilePatternToChunks.
+
+    TODO(#29): Currently, `MergeDim`s are not supported.
+
+    Args:
+      pattern: a `FilePattern` describing a dataset.
+      chunks: split each open dataset into smaller chunks. If not set, the
+        transform will return one file per chunk.
+      local_copy: allow creating local copies of data found in the file pattern.
+      xarray_open_kwargs: keyword arguments to pass to `xarray.open_dataset()`.
+    """
+    self.pattern = pattern
+    self.chunks = chunks
+    self.local_copy = local_copy
+    self.xarray_open_kwargs = xarray_open_kwargs or {}
+    self._max_size_idx = {}
+
+    if pattern.merge_dims:
+      raise ValueError("patterns with `MergeDim`s are not supported.")
+
+  @contextlib.contextmanager
+  def _open_dataset(self, path: str) -> xarray.Dataset:
+    """Open as an XArray Dataset, sometimes with local caching."""
+    with FileSystems().open(path) as file:
+      try:
+        yield xarray.open_dataset(file, **self.xarray_open_kwargs)
+      except (TypeError, OSError) as e:
+
+        if not self.local_copy:
+          raise ValueError(f'cannot open {path!r} with buffering.') from e
+
+        # The cfgrib engine (and others) may fail with the FileSystems method of
+        # opening with BufferedReaders. Here, we open the data locally to make
+        # it easier to work with XArray.
+        with fsspec.open_local(
+            f"simplecache::{path}",
+            simplecache={'cache_storage': '/tmp/files'}
+        ) as fs_file:
+          yield xarray.open_dataset(fs_file, **self.xarray_open_kwargs)
```

Comment:
Rather than using `local_copy` as a fall-back, can we just use an `if` statement?
```suggestion
      if self.local_copy:
        # The cfgrib engine (and others) may fail with the FileSystems method of
        # opening with BufferedReaders. Here, we open the data locally to make
        # it easier to work with XArray.
        with fsspec.open_local(
            f"simplecache::{path}",
            simplecache={'cache_storage': '/tmp/files'}
        ) as fs_file:
          yield xarray.open_dataset(fs_file, **self.xarray_open_kwargs)
      else:
        yield xarray.open_dataset(file, **self.xarray_open_kwargs)
```